### ISSUE TITLE
[Performance Issue] DarkReader slowing down DuckDuckGo rendering in Chrome (recent change)

### ISSUE THREAD

**Issue by bloodgain** [darkreader/darkreader#11047]
### Prerequisites

- [X] I [searched for any existing report](https://github.com/darkreader/darkreader/issues?q=is%3Aissue) about this performance issue to avoid opening a duplicate.
- [X] I have restarted the web browser to check if that resolved this performance issue.
- [X] I can reproduce this performance issue in a new, unmodified web browser profile with Dark Reader installed as the only extension.

### Performance Issue Description

When Dark Reader is active for duckduckgo.com in Chrome Browser, a new Duck Duck Go search results page takes 8-10 seconds to render. When DR is disabled for the site, the page renders in ~1 second. The main Duck Duck Go page also takes slightly longer to render, but only an extra second or so.

How the search results page is accessed does not matter -- searching from address/omni bar, DDG's search box, or clicking a "suggested search" link in DDG. Once the results page is loaded, controls such as loading "more results" or the search box autofill are not delayed.

On further testing, this seems to only effect the render when the Dark or Terminal themes are applied. Light themes do not result in slower renders, or at least not as slow.

### Website Address

duckduckgo.com

### Steps To Reproduce

1. Ensure Dark Reader is active for duckduckgo.com
2. Go to duckduckgo.com
3. If the Dark theme is not selected by default, click the hamburger menu, select "Themes", select "Appearance", and choose the "Dark" or "Terminal" theme.
4. Enter any set of keywords into the Duck Duck Go search bar and hit Enter

### Performance Profile

https://gist.github.com/bloodgain/03bfd3fe5de3647f9e5f221adb7d768e

### Operating System

Windows 10

### Web Browser name and version

Chrome Version 111.0.5563.147 (Official Build) (64-bit)

### Dark Reader version

4.9.62

### Does this happen when Dark Reader is disabled?

No

### Does this happen in other web browsers?

No

### Additional Context

This performance change has occurred within the last few weeks. It was noticed on multiple computers, multiple networks. Until this point, user was unaware that Duck Duck Go was even loading the Dark theme by default. Persisted after history cleared and browser updated. Root cause discovered after checking Duck Duck Go performance on Firefox, where Dark Reader was not installed.

**Comment by ieatacid** [comment]
Same here with Firefox 111.0.1 (64-bit) on Windows 10.

**Comment by RVFET** [comment]
It's almost impossible to stay sane and productive with this extension enabled lately. I think I'm sadly moving on without Dark Reader even tho I've been using it for such a long time...

**Comment by daviddesmet** [comment]
Same here with Edge 111.0.1661.62 (Official build) (arm64)

**Comment by Chaucer85** [comment]
Recreated issue on Microsoft Edge Version 111.0.1661.62 (Official build) (64-bit), Chrome Version 112.0.5615.49 (Official Build) (64-bit), and Firefox 111.0.1 (64-bit). Search with Dark Reader on through DuckDuckGo takes 5-6 seconds, off it takes 1-2 seconds.

**Comment by kands-code** [comment]
Same here with Edge 112.0.1722.34 on Linux.

**Comment by LorenAmelang** [comment]
Just in case you need more clues - my Vivaldi Forum post:
https://forum.vivaldi.net/topic/85436/duckduckgo-response-times-like-8-seconds-in-april-2023

My DevTools shows s2137.css loading in ms, no particular item hung for the 8 seconds. 

Problem doesn't show with:
![image](https://user-images.githubusercontent.com/1297090/230753685-aac92bb3-3d8f-4a3d-afe9-ef23d9a56f64.png)
and:
![image](https://user-images.githubusercontent.com/1297090/230753820-aaa53312-3c9f-4ae6-bcb7-54a0febff37e.png)

**Comment by bershanskiy** [comment]
Hi, thanks for the report. Recently we had identified and fixed a regression leading to a memory leak (fix is not released yet). I would appreciate if you could provide more information about this bug to check if it is what you are observing:
1. Tell me about Dark Reader's memory usage
2. Reload Dark Reader an tell me about memory usage after reload

To reload Dark Reader in Chrome:
1. Open chrome://extensions/
3. Find the card with the Dark Reader icon and click on toggle to change it to left/gray position
4. Toggle Dark Reader back on by toggling the toggle to right/blue position

To measure Dark Reader memory in Chrome:
1. Open Chrome's built-in Task Manager (click "Shift+Esc" or click three dots -> More tools -> Task Manager
2. Find row called "Extension: Dark Reader" and tell me the value from "Memory footprint" column

To reload Dark Reader in Firefox:
1. Open about:addons
2. Find the card with the Dark Reader icon and click on toggle to change it to left/gray position
3. Toggle Dark Reader back on by toggling the toggle to right/blue position

To measure Dark Reader memory in Firefox:
1. Open about:performance
2. Find row called "Dark Reader (addon@darkreader.org)" and tell me the value from "Memory" column

**Comment by Chaucer85** [comment]
> Hi, thanks for the report. Recently we had identified and fixed a regression leading to a memory leak (fix is not released yet). I would appreciate if you could provide more information about this bug to check if it is what you are observing:
> 
>     1. Tell me about Dark Reader's memory usage
> 
>     2. Measure amount of wasted memory in page
> 
>     3. Reload Dark Reader an tell me about memory usage (both 1 and 2) after reload
> 
> 

Firefox memory usage: 5.2 MB
Wasted memory: 2255
Firefox memory usage (after reload): 2.2 MB

Chrome memory usage: 29,296K
Wasted Memory: 1765
Chrome memory usage (after reload): 32,528K

**Comment by bershanskiy** [comment]
Thank you everyone, I identified the problem and prepared a fix. It should be included in 4.9.63 which we plan to release next week.

**Comment by LorenAmelang** [comment]
Tell me about Dark Reader's memory usage
55224K (Vivaldi - Chromium)

Measure amount of wasted memory in page
15549 (This GitHub page)

DevTools failed to load source map: Could not load content for chrome-extension://dfhgmnfclbebfobmblelddiejjcijbjm/lib/browser-polyfill.js.map: System error: net::ERR_BLOCKED_BY_CLIENT

11822 (Slow Duck page)

The error ID:

Trilium Web Clipper
1.0.1
Save web clippings to Trilium Notes.
ID: dfhgmnfclbebfobmblelddiejjcijbjm



Reload Dark Reader an tell me about memory usage (both 1 and 2) after reload

37200K

```
https://github.com/darkreader/darkreader/issue_comments/1501184231/edit_form?textarea_id=issuecomment-1501184231-body&comment_context= net::ERR_ABORTED 403
window.IncludeFragmentElement.fetch @ include-fragment-element-hacks.ts:5
u @ index.js:184
await in u (async)
a @ index.js:162
o @ index.js:131
attributeChangedCallback @ index.js:64
j @ edit.ts:77
capture @ edit.ts:26
c @ index.js:179
await in c (async)
u @ index.js:183
a @ index.js:162
o @ index.js:131
attributeChangedCallback @ index.js:59
r @ index.js:82
a.preload.n.once @ index.js:40
11047:1 Autofocus processing was blocked because a document already has a focused element.
document.querySelector('.darkreader--override').innerText.length
VM110:1 Uncaught TypeError: Cannot read properties of null (reading 'innerText')
    at <anonymous>:1:48
```
(That's for this GitHub page)

```
document.querySelector('.darkreader--override').innerText.length
VM126:1 Uncaught TypeError: Cannot read properties of null (reading 'innerText')
    at <anonymous>:1:48
```
(For the Duck search page)

Reloaded the Duck Search page:

```
Error with Permissions-Policy header: Origin trial controlled feature not enabled: 'interest-cohort'.

DevTools failed to load source map: Could not load content for chrome-extension://dfhgmnfclbebfobmblelddiejjcijbjm/lib/browser-polyfill.js.map: System error: net::ERR_BLOCKED_BY_CLIENT
document.querySelector('.darkreader--override').innerText.length
2121
```

Hope that is what you need...

**Comment by bloodgain** [comment]
@bershanskiy
Behavior still occurs on **Dark Reader v4.9.63** using **Chrome Version 111.0.5563.148 (Official Build) (64-bit)**.

Please let me know if there is additional data I can gather to assist in debugging. I didn't have a way to check if the fix for this was in the latest version.

Side note: The new configurable site toggling successfully detects if DuckDuckGo is using a dark theme (Dark or Terminal) and disables Dark Reader, negating the issue for users who utilize this feature. I highly recommend most users enable this feature.

**Comment by jonahclarsen** [comment]
@bloodgain Based on the PR #11082, looks like this bug was just fixed 3-5 hours ago, while v4.9.63 came out a few days ago. So it will almost certainly be in the _next_ version.

**Comment by bershanskiy** [comment]
@bloodgain @jonahclarsen Thanks for participation, basically:
1. For the time being, I recommend using DuckDuckGo's native dark theme. To do that:
  - Go to DuckDuckGo search result page
  - Open Dark Reader popup and click on block which says `duckduckgo.com` in it
  - Click "Settings", then under "Appearance" click black blob of stuff
  - Reload the page because apparently the page has a bug in dynamically changing color scheme
![image](https://user-images.githubusercontent.com/45960703/231500843-d61eb548-288d-4443-a0a2-9418629888ad.png)

3. I tested older version of Dark Reader on current DuckDuckGo site and it had the same latency. I believe the latency increase is a result of recent DuckDuckGo change which triggers a limitation in Dark Reader (see 4)
4. The PR #11082 is not in current release, but will be in the next one. It helps reduce latency (by making some event listeners passive and thus saturating event loop a tiny bit better), but it does not 
5. Dark Reader currently processes the entire page in essentially a single step (task). This creates a theme for a site in a very fast manner, not delayed by, e.g., ads. Unfortunately, occasionally on larger pages this leads to page load delay. I will try to add a way to break up Dark Reader's work into chunks or add caching so that Dark Reader does not delay these pages.

**Comment by bwc69151** [comment]
I am experiencing this same issue on firefox and I never thought this extension was the cause.

**Comment by FingerlessGlov3s** [comment]
I can't believe the issue was Dark Reader as well!

I had the issue on Firefox on Windows and Brave on Fedora.

**Comment by PolicyPuma4** [comment]
I thought I was going mad. I am also having this issue on Firefox for Android. DarkReader ads numerous seconds to the page loading.

**Comment by GoldenGlimpse** [comment]
I've added https://duckduckgo.com to the Site list (not invert listed) in DarkReader, which does seem to solve my slow loading problem.

**Comment by Matiasgroen** [comment]
I thought I was going insane. I've had this issue for several months on both mobile and desktop and I've only just now found out about it

**Comment by LorenAmelang** [comment]
@ForeverNooob The 'x' in the left side URL box means DR is disabled for that site. Click there, get a checkmark instead, and then On/Off should work. BTW, "disabled" is more inactive than "off"...

**Comment by alexanderby** [comment]
Fixed by #11575
